Rework event bubbling and capture (#20013)
mdn/content: files/en-us/learn/javascript/building_blocks/events/index.md

This behavior can be useful and can also cause unexpected problems. In the next sections we'll see a problem that it causes, and find the solution.

### Video player example

Open up the [show-video-box.html](https://mdn.github.io/learning-area/javascript/building-blocks/events/show-video-box.html) example in a new tab (and the [source code](https://github.com/mdn/learning-area/blob/main/javascript/building-blocks/events/show-video-box.html) in another tab.) It is also available live below:
In this example our page contains a video, which is hidden initially, and a button labeled "Display video". We want the following interaction:

{{ EmbedLiveSample('Video_player_example', '100%', 500, "", "") }}
- When the user clicks the "Display video" button, show the box containing the video, but don't start playing the video yet.
- When the user clicks on the video, start playing the video.
- When the user clicks anywhere in the box outside the video, hide the box.

This example shows and hides a {{htmlelement("div")}} with a {{htmlelement("video")}} element inside it:
The HTML looks like this:

```html
<button>Display video</button>

<div class="hidden">
  <video>

>     <source src="https://raw.githubusercontent.com/mdn/learning-area/master/javascript/building-blocks/events/rabbit320.mp4" type="video/mp4">

Author: > Why removing this line?

Just to keep the example code as simple as possible.

> Also is the url correct? Isn't it possible to have only `rabbit320.mp4` (or `.webm`) like the link two lines below?

Unfortunately not: https://github.com/mdn/yari/issues/5727

- a `<video>` element nested inside the `<div>` element.

 We're using CSS to hide elements with the `"hidden"` class set.

```css hidden
div {
  width: 100%;
  height: 100%;
  background-color: #eee;
}

.hidden {
  display: none;
}

div video {
  padding: 40px;
  display: block;
  width: 400px;
  margin: 40px auto;
}
```

The JavaScript looks like this:

```js
const btn = document.querySelector('button');
const box = document.querySelector('div');
const video = document.querySelector('video');

btn.addEventListener('click', () => box.classList.remove('hidden'));
video.addEventListener('click', () => video.play());
box.addEventListener('click', () => box.classList.add('hidden'));
```

This adds three `'click'` event listeners:

- one on the `<button>`, which shows the `<div>` that contains the `<video>`
- one on the `<video>`, which starts playing the video
- one on the `<div>`, which hides the video

Let's see how this works:

{{ EmbedLiveSample('Video_player_example', '100%', 500) }}

You should see that when you click the button, the box and the video it contains are shown. But then when you click the video, the video starts to play, but the box is hidden again!

> This is because the video is inside the `<div>` — it is part of it — so clicking the video actually runs _both_ the event handlers.

Author: I think "both the event handlers" is fine, but either of the other two are fine as well. Of the two I prefer "both event handlers".

> This is because the video is inside the `<div>` — it is part of it — so clicking the video actually runs _both_ the event handlers.

Author: ```suggestion
The video is inside the `<div>` — it is part of it — so clicking the video runs _both_ the event handlers, causing this behavior.
```

...I think would be better, otherwise "running the event handlers" and "causing the behavior" could be two separate things.

### Fixing the problem with stopPropagation()

> This can be a very annoying behavior, but there is a way to prevent it!

Author: ```suggestion
As we saw in the last section, event bubbling can sometimes create problems, but there is a way to prevent it.
```

?

> The standard [`Event`](/en-US/docs/Web/API/Event) object has a function available on it called [`stopPropagation()`](/en-US/docs/Web/API/Event/stopPropagation) which, when invoked on a handler's event object, makes it so that the first handler is run but the event doesn't bubble any further up the chain, so no more handlers will be run.

Author: I don't think "Still" is right here: it implies that the primary point of `stopPropagation` would be the first part, and the part after "Still" is secondary, which is backwards. I guess you want to split this sentence?

```suggestion
The [`Event`](/en-US/docs/Web/API/Event) object has a function available on it called [`stopPropagation()`](/en-US/docs/Web/API/Event/stopPropagation) which, when called inside an event handler, prevents the event from bubbling up to any other elements.
```

Event capture is disabled by default. To enable it you have to pass the `capture` option in `addEventListener()`.

This example is just like the [bubbling example](#Bubbling_example) we saw earlier, except that we have used the `capture` option:

```html
<body>
  <div id="container">
    <button>Click me!</button>
  </div>
  <pre id="output"></pre>
</body>
```

```js
const output = document.querySelector('#output');
function handleClick(e) {
  output.textContent += `You clicked on a ${e.currentTarget.tagName} element\n`;
}

const container = document.querySelector('#container');
const button = document.querySelector('button');

document.body.addEventListener('click', handleClick, { capture: true });
container.addEventListener('click', handleClick, { capture: true });
button.addEventListener('click', handleClick);
```

{{ EmbedLiveSample('Event capture', '100%', 200, "", "") }}

> In this case, the order of messages is reversed: the `<body>` event handler fires first, followed by the `<div>` event handler, followed by the `<button>` event handler:

Author: I don't think this change is an improvement. "fires first, and next by" doesn't make sense.

```
You clicked on a BODY element
You clicked on a DIV element
You clicked on a BUTTON element
```

> Why bother with both capturing and bubbling? Well, in the bad old days when browsers were much less cross-compatible than they are now, Netscape only used event capturing, and Internet Explorer used only event bubbling. When the W3C decided to try to standardize the behavior and reach a consensus, they ended up with this system that included both, which is the one modern browsers implemented.

Reviewer: ```suggestion
Why bother with both capturing and bubbling? In the bad old days, when browsers were much less cross-compatible than now, Netscape only used event capturing, and Internet Explorer used only event bubbling. When the W3C decided to try to standardize the behavior and reach a consensus, they ended up with this system that included both, which is what modern browsers implement.
```
